src/parser/datetime.rs

```diff
@@ -104,14 +194,115 @@ fn potential_interval_tokens(from: &DateTimeField) -> Vec<IntervalToken> {
     all_toks[offset..].to_vec()
 }
 
-#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+fn build_timezone_offset_second(
+    tokens: &Vec<IntervalToken>,
+    value: &str,
+) -> Result<i64, ParserError> {
+    use IntervalToken::*;
+    let all_formats = [
+        vec![Plus, Num(0), Colon, Num(0)],
+        vec![Dash, Num(0), Colon, Num(0)],
+        vec![Plus, Num(0), Num(0)],
+        vec![Dash, Num(0), Num(0)],
+        vec![Plus, Num(0)],
+        vec![Dash, Num(0)],
+        vec![TzName("".to_string())],
+        vec![Zulu],
+    ];
+
+    let mut is_positive = true;
+    let mut hour_offset: Option<i64> = None;
+    let mut minute_offset: Option<i64> = None;
+
+    for format in all_formats.iter() {
+        let actual = tokens.iter();
+
+        if actual.len() != format.len() {
+            continue;
+        }
+
+        for (i, (atok, etok)) in actual.zip(format).enumerate() {
+            match (atok, etok) {
+                (Colon, Colon) | (Plus, Plus) => { /* Matching punctuation */ }
+                (Dash, Dash) => {
+                    is_positive = false;
+                }
+                (Num(val), Num(_)) => {
+                    let val = *val;
+                    if hour_offset.is_none() {
+                        // We haven't seen an hour offset yet so lets try to parse that
+                        if val <= 24 {
+                            hour_offset = Some(val as i64);
+                        } else {
+                            return Err(ParserError::ParserError(format!(
+                                "Error parsing timezone string ({}): timezone hour invalid {}",
+                                value, val
+                            )));
+                        }
+                    } else if minute_offset.is_none() {
+                        if val <= 60 {
+                            minute_offset = Some(val as i64);
+                        } else {
+                            return Err(ParserError::ParserError(format!(
+                                "Error parsing timezone string ({}): timezone minute invalid {}",
+                                value, val
+                            )));
+                        }
+                    } else {
+                        // We've already seen an hour and a minute so we should never see another number
+                        return Err(ParserError::ParserError(format!(
+                            "Error parsing timezone string ({}): invalid value {} at token index {}", value,
+                            val, i
+                        )));
+                    }
```

Comment:
This is not a request for change, just pointing out another possible idiom that
I prefer to use: it is possible to use `match` expressions to both show that
you are building up the two variables, and to make it impossible to not handle
one of the error cases:

```
let val = *val;
match (hour_offset, minute_offset) {
    (None, None) => if val <= 24 {
         hour_offset = Some(val as i64);
     } else {
         return Err(ParserError::ParserError(format!(
             "Error parsing timezone string ({}): timezone hour invalid {}",
             value, val
         )));
     },
     (Some(_), None) => if val <= 60 {
         minute_offset = Some(val as i64);
     } else {
         return Err(ParserError::ParserError(format!(
             "Error parsing timezone string ({}): timezone minute invalid {}",
             value, val
         )));
     },
     // We've already seen an hour and a minute so we should never see another number
     (Some(_), Some(_)) => return Err(ParserError::ParserError(format!(
         "Error parsing timezone string ({}): invalid value {} at token index {}", value,
         val, i
     ))),
     (None, Some(_)) => unreachable!("parsed a minute before an hour"),
}
```

Reply:
this is absolutely better